# Town Council Budget Dataset

In [3]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()  # for plot styling
import numpy as np
import pandas as pd


In [170]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


In [4]:
data_original = pd.read_csv('Pressupostos_dels_ens_municipals_de_Catalunya.csv')

In [5]:
data_original.shape

(822901, 11)

In [6]:
list(data_original)

['ANY_EXERCICI',
 'TIPUS_PARTIDA',
 'TIPUS_CLASSIF',
 'CODI_PANTALLA',
 'NIVELL',
 'DESCRIPCIO',
 'IMPORT',
 'ID_ESTRUCTURA',
 'ESTRUCTURA',
 'CODI_ENS',
 'NOM_COMPLERT']

In [7]:
set(data_original['ANY_EXERCICI'])

{'01/01/2016 12:00:00 AM', '01/01/2017 12:00:00 AM'}

In [8]:
set(data_original['NOM_COMPLERT'])

{'Ajuntament de Roda de Ter',
 'Ajuntament del Prat de Llobregat',
 "Ajuntament de Vilanova de l'Aguda",
 "Institut Municipal d'Esports (IME)",
 'Ajuntament de Sant Ramon',
 'Ajuntament de Vielha e Mijaran',
 'Grup de Serveis Sanitaris i Socials Sant Roc',
 'Ajuntament de Bonastre',
 "Ajuntament d'Almacelles",
 'Fundació Pública per a la prestació del servei de Museu Municipal',
 'Ajuntament de Sanaüja',
 'Ajuntament de Santa Oliva',
 'Ajuntament de Torroella de Fluvià',
 'Ajuntament de Vila-rodona',
 "Patronat Municipal d'Esports de Ripollet",
 'Museu de l\'Aquarel·la "J.  Martínez Lozano"',
 'Ajuntament de Soses',
 'Ajuntament de Campelles',
 'Ajuntament de la Nou de Berguedà',
 "Institut Municipal d'Esports i de Lleure de Montcada i Reixac",
 'Ajuntament de Sant Julià de Vilatorta',
 "Entitat Municipal Descentralitzada d'Arestui",
 "OAL Institut Municipal d'Educació",
 'Ajuntament de Sant Pau de Segúries',
 'Ajuntament de Tivenys',
 'Ajuntament de Sant Cugat Sesgarrigues',
 'Ajuntam

In [199]:
data_original

ANY_EXERCICI TIPUS_PARTIDA TIPUS_CLASSIF  CODI_PANTALLA  \
0       01/01/2016 12:00:00 AM             I             E           3.00   
1       01/01/2016 12:00:00 AM             I             E          36.00   
2       01/01/2016 12:00:00 AM             I             E          38.00   
3       01/01/2016 12:00:00 AM             I             E         389.00   
4       01/01/2016 12:00:00 AM             I             E          39.00   
5       01/01/2016 12:00:00 AM             I             E         399.00   
6       01/01/2016 12:00:00 AM             I             E           4.00   
7       01/01/2016 12:00:00 AM             I             E          42.00   
8       01/01/2016 12:00:00 AM             I             E         420.00   
9       01/01/2016 12:00:00 AM             I             E         420.90   
10      01/01/2016 12:00:00 AM             I             E          45.00   
11      01/01/2016 12:00:00 AM             I             E         450.00   
12      01/01/2016 12:00:00 AM             I             E         450.00   
13      01/01/2016 12:00:00 AM             I             E          46.00   
14      01/01/2016 12:00:00 AM             I             E         462.00   
15      01/01/2016 12:00:00 AM             I             E         464.00   
16      01/01/2016 12:00:00 AM             I             E         467.00   
17      01/01/2016 12:00:00 AM             I             E          49.00   
18      01/01/2016 12:00:00 AM             I             E         490.00   
19      01/01/2016 12:00:00 AM             I             E           5.00   
20      01/01/2016 12:00:00 AM             I             E          52.00   
21      01/01/2016 12:00:00 AM             I             E           6.00   
22      01/01/2016 12:00:00 AM             I             E          61.00   
23      01/01/2016 12:00:00 AM             I             E         619.00   
24      01/01/2016 12:00:00 AM             I             E           7.00   
25      01/01/2016 12:00:00 AM             I             E          70.00   
26      01/01/2016 12:00:00 AM             I             E          72.00   
27      01/01/2016 12:00:00 AM             I             E         723.00   
28      01/01/2016 12:00:00 AM             I             E         723.90   
29      01/01/2016 12:00:00 AM             I             E          75.00   
30      01/01/2016 12:00:00 AM             I             E         753.00   
31      01/01/2016 12:00:00 AM             I             E          76.00   
32      01/01/2016 12:00:00 AM             I             E         762.00   
33      01/01/2016 12:00:00 AM             D             F           0.00   
34      01/01/2016 12:00:00 AM             D             F           1.00   
35      01/01/2016 12:00:00 AM             D             F          11.00   
36      01/01/2016 12:00:00 AM             D             F           1.00   
37      01/01/2016 12:00:00 AM             D             F          13.00   
38      01/01/2016 12:00:00 AM             D             F         130.00   
39      01/01/2016 12:00:00 AM             D             F         132.00   
40      01/01/2016 12:00:00 AM             D             F         133.00   
41      01/01/2016 12:00:00 AM             D             F         134.00   
42      01/01/2016 12:00:00 AM             D             F         135.00   
43      01/01/2016 12:00:00 AM             D             F         136.00   
44      01/01/2016 12:00:00 AM             D             F          15.00   
45      01/01/2016 12:00:00 AM             D             F         150.00   
46      01/01/2016 12:00:00 AM             D             F         151.00   
47      01/01/2016 12:00:00 AM             D             F         152.00   
48      01/01/2016 12:00:00 AM             D             F        1521.00   
49      01/01/2016 12:00:00 AM             D             F        1522.00   
50      01/01/2016 12:00:00 AM             D             F         153.00   
51      01/01/

In [9]:
print(len(set(data_original.NOM_COMPLERT)))

1345


In [10]:
data_16 = data_original.loc[data_original.ANY_EXERCICI == '01/01/2016 12:00:00 AM']
data_16.shape

(624092, 11)

In [11]:
set(data_16['ANY_EXERCICI'])

{'01/01/2016 12:00:00 AM'}

In [12]:
data_16_aj = data_16.loc[data_16.NOM_COMPLERT.str.contains('Ajuntament')]
data_16_aj.shape

(434340, 11)

In [13]:
name, counts = np.unique(data_16_aj.NOM_COMPLERT, return_counts = True)
dict(zip(name,counts))

{"Ajuntament d'Abrera": 317,
 "Ajuntament d'Agramunt": 832,
 "Ajuntament d'Aguilar de Segarra": 347,
 "Ajuntament d'Agullana": 251,
 "Ajuntament d'Aiguafreda": 275,
 "Ajuntament d'Aiguamúrcia": 830,
 "Ajuntament d'Aiguaviva": 259,
 "Ajuntament d'Aitona": 833,
 "Ajuntament d'Albanyà": 241,
 "Ajuntament d'Albatàrrec": 243,
 "Ajuntament d'Albesa": 832,
 "Ajuntament d'Albinyana": 833,
 "Ajuntament d'Albons": 242,
 "Ajuntament d'Alcanar": 830,
 "Ajuntament d'Alcanó": 833,
 "Ajuntament d'Alcarràs": 834,
 "Ajuntament d'Alcoletge": 833,
 "Ajuntament d'Alcover": 332,
 "Ajuntament d'Aldover": 833,
 "Ajuntament d'Alella": 343,
 "Ajuntament d'Alfara de Carles": 249,
 "Ajuntament d'Alfarràs": 833,
 "Ajuntament d'Alforja": 833,
 "Ajuntament d'Alfés": 832,
 "Ajuntament d'Algerri": 833,
 "Ajuntament d'Alguaire": 833,
 "Ajuntament d'Alió": 255,
 "Ajuntament d'Almacelles": 833,
 "Ajuntament d'Almatret": 833,
 "Ajuntament d'Almenar": 833,
 "Ajuntament d'Almoster": 302,
 "Ajuntament d'Alp": 832,
 "Ajuntam

In [14]:
print(len(name))

882


In [15]:
set(data_16_aj.loc[data_16_aj.NOM_COMPLERT == 'Ajuntament de Mieres'].DESCRIPCIO).issubset(set(data_16_aj.loc[data_16_aj.NOM_COMPLERT == 'Ajuntament de Menàrguens'].DESCRIPCIO))

True

In [16]:
data_16_aj_no0=data_16_aj.loc[data_16_aj.IMPORT > 0.0]

In [17]:
name, counts = np.unique(data_16_aj_no0.NOM_COMPLERT, return_counts = True)
dict(zip(name,counts))

{"Ajuntament d'Abrera": 249,
 "Ajuntament d'Agramunt": 230,
 "Ajuntament d'Aguilar de Segarra": 257,
 "Ajuntament d'Agullana": 157,
 "Ajuntament d'Aiguafreda": 205,
 "Ajuntament d'Aiguamúrcia": 204,
 "Ajuntament d'Aiguaviva": 172,
 "Ajuntament d'Aitona": 270,
 "Ajuntament d'Albanyà": 122,
 "Ajuntament d'Albatàrrec": 114,
 "Ajuntament d'Albesa": 209,
 "Ajuntament d'Albinyana": 218,
 "Ajuntament d'Albons": 147,
 "Ajuntament d'Alcanar": 270,
 "Ajuntament d'Alcanó": 122,
 "Ajuntament d'Alcarràs": 256,
 "Ajuntament d'Alcoletge": 235,
 "Ajuntament d'Alcover": 287,
 "Ajuntament d'Aldover": 212,
 "Ajuntament d'Alella": 299,
 "Ajuntament d'Alfara de Carles": 148,
 "Ajuntament d'Alfarràs": 280,
 "Ajuntament d'Alforja": 187,
 "Ajuntament d'Alfés": 168,
 "Ajuntament d'Algerri": 192,
 "Ajuntament d'Alguaire": 238,
 "Ajuntament d'Alió": 161,
 "Ajuntament d'Almacelles": 243,
 "Ajuntament d'Almatret": 164,
 "Ajuntament d'Almenar": 210,
 "Ajuntament d'Almoster": 192,
 "Ajuntament d'Alp": 220,
 "Ajuntam

In [18]:
print(np.amax(counts), np.amin(counts), np.mean(counts), np.median(counts))

387 47 201.99092970521542 193.0


In [19]:
name.shape

(882,)

In [20]:
np.unique(data_16_aj_no0.TIPUS_PARTIDA)

array(['D', 'I'], dtype=object)

In [21]:
np.unique(data_16_aj_no0.TIPUS_CLASSIF)

array(['E', 'F'], dtype=object)

In [22]:
len(np.unique(data_16_aj_no0.CODI_PANTALLA))

498

In [23]:
np.unique(data_16_aj_no0.NIVELL)

array([1, 2, 3, 4])

In [24]:
np.unique(data_16_aj_no0.ID_ESTRUCTURA)

array([7])

In [25]:
len(np.unique(data_16_aj_no0.ESTRUCTURA))

532

In [26]:
len(np.unique(data_16_aj_no0.DESCRIPCIO))

591

# Explanation
The maximum number of rows an Ajuntament has is 387. The total number of descriptions (budget concept) is 591 meaning that not all ajuntaments have the same concepts which is bad...

Now we have to discover if there is a structure (hierarchy) in the concepts

In [27]:
np.sum(counts < 100)

7

In [28]:
np.sum(counts < 150)

171

In [29]:
np.sum(counts > 350)

5

In [32]:
def description_by_level_despesa (data, level):
    desc, count = np.unique(data.loc[(data.NIVELL == level) & (data.TIPUS_PARTIDA == 'D') & (data.TIPUS_CLASSIF == 'F'), ['DESCRIPCIO']], return_counts = True)
    return dict(zip(desc,counts))

res = description_by_level_despesa(data_16_aj_no0, 1)
print("length: ", len(res))
res

length:  6


{'Actuacions de caràcter econòmic': 249,
 'Actuacions de caràcter general': 230,
 'Actuacions de protecció i promoció social': 257,
 'Deute públic': 157,
 'Producció de béns públics de caràcter preferent': 205,
 'Serveis públics bàsics': 204}

In [123]:
res = description_by_level_despesa(data_16_aj_no0, 2)
print("length: ", len(res))
res

length:  24


{'Administració financera i tributària': 249,
 'Agricultura, ramaderia i pesca': 230,
 'Altres actuacions de caràcter econòmic': 257,
 "Altres prestacions econòmiques a favor d'empleats": 157,
 'Benestar comunitari ': 205,
 'Comerç, turisme i petites i mitjanes empreses': 204,
 'Cultura': 172,
 'Deute públic': 270,
 'Educació': 122,
 'Esport': 114,
 "Foment de l'ocupació": 209,
 'Habitatge i urbanisme': 218,
 'Indústria i energia': 147,
 'Infraestructures': 270,
 'Investigació, desenvolupament i innovació': 122,
 'Medi ambient': 256,
 'Pensions': 235,
 'Sanitat': 287,
 'Seguretat i mobilitat ciutadana': 212,
 'Serveis de caràcter general': 299,
 'Serveis socials i promoció social': 148,
 'Transferències a altres administracions públiques': 280,
 'Transport públic': 187,
 'Òrgans de govern': 168}

In [124]:
res = description_by_level_despesa(data_16_aj_no0, 3)
print("length: ", len(res))
res

length:  87


{"Abastament domiciliari d'aigua potable": 249,
 'Administració general': 230,
 "Administració general d'educació": 257,
 "Administració general d'esports": 157,
 "Administració general d'habitatge i urbanisme": 205,
 "Administració general d'indústria i energia": 204,
 "Administració general d'infraestructures": 172,
 'Administració general de comerç, turisme i pimes': 270,
 'Administració general de cultura': 122,
 'Administració general del medi ambient': 114,
 'Administració general del transport': 209,
 'Administració gral. agricultura, ramaderia i pesca': 218,
 'Administració gral. de seguretat i protecció civil': 147,
 'Altres actuacions agricultura, ramaderia i pesca': 270,
 'Altres actuacions sectorials': 122,
 'Altres infraestructures': 256,
 "Altres prestacions econòmiques a favor d'empleats": 235,
 'Assistència social primària': 287,
 'Atenció als ciutadans': 212,
 'Biblioteques i arxius': 299,
 'Camins veïnals': 148,
 'Carreteres': 280,
 'Cementiri i serveis funeraris.': 1

## Concepts

The following function shows the spendings of a given (by name) town council sorted by "codi pantalla". You can observe that there is a pattern. The more digits there are the deeper is the level (NIVELL).

In [173]:
def concepte_despesa_ajuntament(data, ajuntament):
    tmp = data.loc[(data.TIPUS_PARTIDA == 'D') & (data.NOM_COMPLERT == ajuntament),['DESCRIPCIO', 'IMPORT', 'CODI_PANTALLA', 'NIVELL']]
    tmp[['CODI_PANTALLA']] = tmp[['CODI_PANTALLA']].astype(str)
    return tmp.sort_values(['CODI_PANTALLA'])
    
pd.set_option('display.max_rows', 500)
concepte_despesa_ajuntament(data_16_aj_no0, "Ajuntament d'Ogassa")

DESCRIPCIO    IMPORT  \
44869                                       Deute públic   46215.0   
44872                             Serveis públics bàsics   88700.0   
45003                               Despeses de personal   95000.0   
44870                                       Deute públic   46215.0   
44871                                       Deute públic   46215.0   
45004                                   Personal laboral   72000.0   
45005                                        Laboral fix   72000.0   
45006                              Retribucions bàsiques   72000.0   
44880                              Habitatge i urbanisme   77700.0   
44882     Urbanisme: Planejament,gest exec. i disc. urb.   77700.0   
44889                               Benestar comunitari    11000.0   
45007  Quotes, prestac. i desp. socials a/c de l'ocup...   23000.0   
45008                                     Quotes socials   23000.0   
45009                                   Seguretat Social   23000.0   
44891             Abastament domiciliari d'aigua potable   11000.0   
45010                Despeses corrents en béns i serveis  161485.0   
45011             Reparacions, manteniment i conservació   66000.0   
45012                   Infraestructures i béns naturals   13000.0   
45013                    Edificis i altres construccions   19000.0   
45014   Maquinària, instal·lacions tècniques i utillatge   33000.0   
45015                Equips per a processos d'informació    1000.0   
45016              Material, subministraments i d'altres   87385.0   
45017                                 Material d'oficina    2000.0   
45018                          Ordinari no inventariable    2000.0   
45020                                  Energia elèctrica   15000.0   
45019                                   Subministraments   22000.0   
45021                                                Gas    7000.0   
45023                       Serveis de telecomunicacions    3000.0   
45022                                      Comunicacions    4900.0   
45024                                            Postals     700.0   
45025                   Altres despeses en comunicacions    1200.0   
45026                              Primes d'assegurances    2500.0   
45027                                            Tributs    1000.0   
45028                Tributs de les comunitats autònomes    1000.0   
45029                                  Despeses diverses   22385.0   
45030        Atencions protocol·làries i representatives    1200.0   
45031                            Publicitat i propaganda    7000.0   
45032                  Activitats culturals i esportives   12000.0   
45033                           Altres despeses diverses    2185.0   
45034  Treballs fets per altres empreses i professionals   32600.0   
45035                                   Neteja i higiene    5500.0   
45036                 Custòdia, dipòsit i emmagatzematge    2000.0   
45037                         Estudis i treballs tècnics    9000.0   
45038         Serveis de recaptació a favor de l'entitat    3000.0   
45039   Altres feines fetes per empreses i professionals   13100.0   
45040                 Indemnitzacions per raó del servei    8100.0   
45042                 Dels membres dels òrgans de govern    2000.0   
45041                                          Locomoció    2100.0   
45043                           Del personal no directiu     100.0   
45044                             Altres indemnitzacions    6000.0   
44913    Producció de béns públics de caràcter preferent   32200.0   
45045                                Despeses financeres   13515.0   
44914                                            Sanitat    5000.0   
45046    De préstecs i altres operacions financeres en €   12615.0   
45047                                         Interessos   12615.0   
44916  Hospitals, serveis assistencials, centres de s...    5000.0   
44917                                           Educació

# Ingressos - despesa

In [143]:
#data_16_aj_no0

def get_import_ajuntament(data, codi_ens, tipus):
    return np.sum(data.loc[(data.CODI_ENS == codi_ens) & (data.TIPUS_PARTIDA == tipus), ['IMPORT']].IMPORT)

def get_ingressos_despesa_per_ajuntament(data):
    codi_ens = np.unique(data.CODI_ENS)
    res = np.empty((codi_ens.shape[0],3))
    
    i = 0;
    for ens in codi_ens:
        ingres = get_import_ajuntament(data, ens, 'I')
        despesa = get_import_ajuntament(data, ens, 'D')
        res[i] = [ens, ingres, despesa]
        i = i +1
        
    return res
        

In [144]:
ingres_despesa = get_ingressos_despesa_per_ajuntament(data_16_aj_no0)


array([[8.00180001e+08, 4.52862159e+07, 9.52842650e+07],
       [8.00230008e+08, 1.05162000e+07, 2.12483121e+07],
       [8.00390004e+08, 4.19815289e+07, 8.85857316e+07],
       ...,
       [4.39057000e+09, 1.27906000e+08, 2.70731853e+08],
       [4.39060000e+09, 1.62978725e+07, 3.23994095e+07],
       [4.39076001e+09, 3.44369000e+07, 7.03371000e+07]])

### Ratio outcome / income
We can observe that the ratio is 1/2 most of the times. Probably meaning that the output is added twice. This confirms what we suspected in the **concepts** section

In [147]:
rati_income_outcome = ingres_despesa[:, 1] / ingres_despesa[:,2]
rati_income_outcome

array([0.47527486, 0.49491931, 0.47390847, 0.47884304, 0.46810369,
       0.471914  , 0.47871747, 0.47988971, 0.47289738, 0.50008066,
       0.48557806, 0.48651019, 0.47254921, 0.51455848, 0.49690417,
       0.46352294, 0.56548065, 0.47506806, 0.50740893, 0.48647117,
       0.46212542, 0.48214916, 0.49935141, 0.47978873, 0.47330138,
       0.51547832, 0.50483833, 0.47687954, 0.47622266, 0.47517577,
       0.49141123, 0.48844239, 0.49503962, 0.47511836, 0.48832893,
       0.49552842, 0.51283907, 0.48448334, 0.47079229, 0.49168559,
       0.4803879 , 0.4952988 , 0.49312309, 0.46287262, 0.47299963,
       0.49092361, 0.51666307, 0.46134055, 0.48272626, 0.51271847,
       0.47629888, 0.49770467, 0.44046898, 0.46595055, 0.50316567,
       0.474611  , 0.49392207, 0.48681057, 0.47107242, 0.47757469,
       0.49226243, 0.4827205 , 0.48963372, 0.49478539, 0.48986378,
       0.47506056, 0.50042672, 0.53183243, 0.49615261, 0.53288403,
       0.49308263, 0.45665938, 0.48088049, 0.48216345, 0.48489

### Spending (outcome) by level and town council

In [197]:
def sum_outcome_ajuntament_level(data, ajuntament, level):
    tmp = data.loc[(data.TIPUS_PARTIDA == 'D') & (data.NOM_COMPLERT == ajuntament) & (data.NIVELL == level),['IMPORT']]
    return np.sum(tmp)

def print_outcome_by_level_ajuntament(data, ajuntament):
    for i in range(1,5):
        print('Level ' , i, ': %06.0f' %sum_outcome_ajuntament_level(data, ajuntament , i))
print_outcome_by_level_ajuntament(data_16_aj_no0, "Ajuntament d'Ogassa")



Level  1 : 730000
Level  2 : 730000
Level  3 : 719000
Level  4 : 181985


In [198]:
print_outcome_by_level_ajuntament(data_16_aj_no0, 'Ajuntament de Barcelona')

Level  1 : 4911868464
Level  2 : 4911868464
Level  3 : 4556627884
Level  4 : 904767762


In [140]:
a[0] = ([1,2,3])
a

array([[1., 2., 3.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [ ]:
a